lib/withOnyx.d.ts

```diff
@@ -0,0 +1,159 @@
+};
+
+/**
+ * Represents the function `key` and `selector` mapping options between an Onyx key and the component's prop.
+ *
+ * The function signature and return type of `selector` must match with the type of the component's prop,
+ * otherwise an error will be thrown.
+ *
+ * @example
+ * ```ts
+ * // Onyx prop with `function` key and selector
+ * onyxProp: {
+ *     key: ({reportId}) => ONYXKEYS.ACCOUNT,
+ *     selector: (value: Account | null) => value?.id ?? '',
+ * },
+ * ```
+ */
+type BaseMappingFunctionKeyAndSelector<
+    TComponentProps,
+    TOnyxProps,
+    TOnyxProp extends keyof TOnyxProps,
+    TOnyxKey extends OnyxKey
+> = {
+    key: (props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey;
+    selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+};
+
+/**
+ * Represents the mapping options between an Onyx key and the component's prop with all its possibilities.
+ */
+type Mapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TOnyxKey extends OnyxKey> = BaseMapping<
+    TComponentProps,
+    TOnyxProps
+> &
+    (
+        | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxEntry<KeyValueMapping[TOnyxKey]>>
+        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
+        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
+    );
+
+/**
+ * Represents the mapping options between an Onyx collection key without suffix and the component's prop with all its possibilities.
+ */
+type CollectionMapping<
+    TComponentProps,
+    TOnyxProps,
+    TOnyxProp extends keyof TOnyxProps,
+    TOnyxKey extends CollectionKeyBase
+> = BaseMapping<TComponentProps, TOnyxProps> &
+    (
+        | BaseMappingKey<
+              TComponentProps,
+              TOnyxProps,
+              TOnyxProp,
+              TOnyxKey,
+              OnyxCollectionEntries<KeyValueMapping[TOnyxKey]>
+          >
+        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
+        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
+    );
+
+/**
+ * Represents an union type of all the possible Onyx key mappings.
+ * Each `OnyxPropMapping` will be associated with its respective Onyx key, ensuring different type-safety for each object.
+ */
+type OnyxPropMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps> = {
+    [TOnyxKey in OnyxKey]: Mapping<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>;
+}[OnyxKey];
+
+/**
+ * Represents an union type of all the possible Onyx collection keys without suffix mappings.
+ * Each `OnyxPropCollectionMapping` will be associated with its respective Onyx key, ensuring different type-safety for each object.
+ */
+type OnyxPropCollectionMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps> = {
+    [TOnyxKey in CollectionKeyBase]: CollectionMapping<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>;
```

Comment:
so the `key` property on an object passed to `withOnyx` only takes normal keys or collection keys without suffix?

It cannot take `report_<some_report_id>` kind of key?

or am I misreading the type definitions here? `CollectionKeyBase` is a union of collection name strings that are NOT defined using template literals right?

Reply:
FYI, [here](https://github.com/blazejkustra/App/pull/2/files) you can have all possibilities. You can download and play around with it 👍 

Correct me if I'm wrong @fabioh8010: `OnyxPropCollectionMapping` is responsible only for collection keys without suffixes (`report_`). `OnyxPropMapping` handled all other keys, including collection keys with suffixes (`report_{id}`)

Reply:
@blazejkustra Yes, that is correct. `OnyxPropMapping` handles both normal keys and collection keys WITH suffix. `OnyxPropCollectionMapping` only handles collection keys WITHOUT suffix because of its different Onyx value.

cc @hayata-suenaga

Reply:
ahhh I didn't realize `OnyxPropMapping` also handles a collection key with suffix thank you for helping me understand 🙇